moveit_ros/moveit_servo/src/servo.cpp

```diff
@@ -649,39 +667,41 @@ std::optional<PoseCommand> Servo::toPlanningFrame(const PoseCommand& command, co
 KinematicState Servo::getCurrentRobotState() const
 {
   moveit::core::RobotStatePtr robot_state = planning_scene_monitor_->getStateMonitor()->getCurrentState();
-  const moveit::core::JointModelGroup* joint_model_group =
-      robot_state->getJointModelGroup(servo_params_.move_group_name);
-  const auto joint_names = joint_model_group->getActiveJointModelNames();
-
-  KinematicState current_state(joint_names.size());
-  current_state.joint_names = joint_names;
-  robot_state->copyJointGroupPositions(joint_model_group, current_state.positions);
-  robot_state->copyJointGroupVelocities(joint_model_group, current_state.velocities);
-  robot_state->copyJointGroupAccelerations(joint_model_group, current_state.accelerations);
-
-  return current_state;
+  return extractRobotState(robot_state, servo_params_.move_group_name);
 }
 
-std::pair<bool, KinematicState> Servo::smoothHalt(const KinematicState& halt_state)
+std::pair<bool, KinematicState> Servo::smoothHalt(const moveit::core::RobotStatePtr& robot_state,
+                                                  const KinematicState& halt_state)
 {
   bool stopped = false;
   auto target_state = halt_state;
-  const KinematicState current_state = getCurrentRobotState();
 
-  const size_t num_joints = current_state.joint_names.size();
-  for (size_t i = 0; i < num_joints; ++i)
+  // Get joint bounds from robot state for current move group.
+  const moveit::core::JointModelGroup* joint_model_group =
+      robot_state->getJointModelGroup(servo_params_.move_group_name);
+  const moveit::core::JointBoundsVector joint_bounds = joint_model_group->getActiveJointModelsBounds();
+
+  // apply scaling to target velocity based on robot's limits
+  Eigen::VectorXd acceleration = -target_state.velocities / servo_params_.publish_period;
+  const double joint_acceleration_limit_scale = jointLimitAccelerationScalingFactor(acceleration, joint_bounds, 1.0);
+  target_state.velocities += joint_acceleration_limit_scale * acceleration * servo_params_.publish_period;
+
```

Comment:
This acceleration scaling is what I was trying to get you to remove with the previous comments. The ones about changing direction and `override_acceleration_scaling_factor`.

```suggestion
```

Reply:
Okay, I see. One problem is that the robot will be commanded to zero velocity immediately if we don't limit the acceleration. I wouldn't call that a smooth hault. The smoothing might help make the stop more gradual. But what benefit is there for gradual slow down via smoothing (e.g. butterworth filtering) vs. acceleration limited? One benefit of acceleration limited is that it will slow down as quickly as possible while not violating the hardware limits.

Reply:
I think @AndyZe's comment about direction switching is valid, and I hadn't thought of that since I was thinking about the full trajectory planning case where scaling down acceleration + scaling up time accelerating works equivalently to the unscaled case, as far as the space swept by the robot motion.

Is there any way to do the same within the current servo framework now that we have the rolling window? Like, time scaling https://www2.ece.ohio-state.edu/~zhang/RoboticsClass/docs/LN9_TrajectoryGeneration.pdf

Reply:
Suggest moving the acceleration smoothing to a new PR so we can merge this one.

Again, I think an acceleration-smoothing plugin in `online_signal_smoothing` is the way to go.

Reply:
@sea-bass I like the idea of generating a path and then time scaling to enforce dynamic constraints. The biggest problem is that a twice-differentiable path is needed. We would need to fill in the path between the waypoints with a smooth function. One option is to use circular arcs like done in [here](https://roboticsproceedings.org/rss08/p27.pdf) section IV. 

For the full trajectory planning case, deviation from the path found by the global planner is detrimental because those are the points that are collision checked, so it could result in a collision. However, since servo is constantly running collision checking, this is not a big concern. Maybe this graphic will help clearify.

![Screenshot from 2023-12-28 10-48-42](https://github.com/ros-planning/moveit2/assets/28868750/3fbc9e3e-9013-4989-85de-17c66528183f)

In this scenario, the blue dots represent the commands the user sends to the robot. The plot on the left shows the user commanding a sharp 90 degree angle. On the right, the same situation is shown but with acceleration limiting applied. Because the command violates the robot's acceleration limits, the commands are modified (shown in red). In this case, the red dots are points that are being collision checked, so colliding due to deviating from the reference should not be an issue. However, if we knew we had the entire trajectory ahead of time, then we could do time parameterization in such as way that no limits are violated and the path is followed exactly.

Does anyone have thoughts in this?

Reply:
> Suggest moving the acceleration smoothing to a new PR so we can merge this one.
> 
> Again, I think an acceleration-smoothing plugin in `online_signal_smoothing` is the way to go.

Maybe that is the best option.